src/App.tsx

```diff
@@ -36,35 +50,63 @@ function App(props: Props) {
       state.signer &&
       state.address &&
       roundId &&
-      (previousAddress === null || previousSigner === null)
+      (!previousAddress || !previousSigner)
     ) {
       const address = state.address;
       const message = `UMA Protocol one time key for round: ${roundId}`;
+      const hexMessage = web3.utils.toHex(message);
       const keyExists = signingKeys[address];
-      if (!keyExists || keyExists.roundMessage !== message) {
-        state.signer
-          .signMessage(message)
-          .then((msg) => {
-            const key = {} as {
-              publicKey: string;
-              privateKey: string;
-              roundMessage: string;
-            };
-
-            const privateKey = derivePrivateKey(msg);
-            const publicKey = recoverPublicKey(privateKey);
-            key.privateKey = privateKey;
-            key.publicKey = publicKey;
-            key.roundMessage = message;
-
-            setSigningKeys((prevKeys) => {
-              return { ...prevKeys, [address]: key };
+      let keysInStorage =
+        localStorage.getItem(SIGNING_KEYS_STORAGE_KEY) || ({} as StorageKeys);
+
+      let keyExistsInStorage = false;
+      if (typeof keysInStorage === "string") {
+        keysInStorage = JSON.parse(keysInStorage) as StorageKeys;
+
+        keyExistsInStorage = has(keysInStorage, `${hexMessage}.${address}`);
+      }
+
+      if (keyExistsInStorage) {
+        const loggedInKey = keysInStorage[hexMessage][address];
+        setSigningKeys((prevKeys) => {
+          return { ...prevKeys, [address]: loggedInKey };
+        });
+      } else {
+        if (!keyExists || keyExists.roundMessage !== message) {
+          state.signer
+            .signMessage(message)
+            .then((msg) => {
+              const key = {} as SigningKey;
+
+              const privateKey = derivePrivateKey(msg);
+              const publicKey = recoverPublicKey(privateKey);
+              key.privateKey = privateKey;
+              key.publicKey = publicKey;
+              key.roundMessage = message;
+              key.roundId = roundId;
+
+              let keysToBackup = {};
+              if (typeof keysInStorage === "string") {
+                const parsedJSON = JSON.parse(keysInStorage);
+                keysToBackup = { ...parsedJSON };
+              } else {
+                keysToBackup = { ...keysInStorage };
+              }
```

Comment:
What is this else case? Can localstorage return an object? When would that happen? Or is this just because we cast the `||` case above to `StorageKeys`?

Reply:
It's a typescript thing. Basically unless I do a type check here, it won't let me spread over a string.